### ISSUE TITLE
catboost encoder get different result with catboost

### ISSUE THREAD

**Issue by ccylance** [scikit-learn-contrib/category_encoders#436]
## Steps to Reproduce the Problem
Hi, I noticed CatBoostEncoder  supports time-aware encoding. However, when I do testing, for the same object in category, CatBoostEncoder return the same result for different rows, while catboost not. 
For example, suppose we have feature named color with value blue, red, green and blue.  
CatBoostEncoder may return the result with 0.2, 0.4, 0.4,0.2. We can see the value blue with get the same value. 
But in catboost it will return 0.2, 0.4, 0.4,0.25.  The value of blue will be affected by the order.

**Comment by bmreiniger** [comment]
Can you provide code with your example?

Note that `fit_transform` should produce different values per category, whereas `transform` should not.

**Comment by ccylance** [comment]
Here are my code
![image](https://github.com/scikit-learn-contrib/category_encoders/assets/54780246/be769118-ca0e-4feb-a7d1-f88e8fd0745a)
As the code showed, different categories will have different results. But in CatBoost, the same category with different orders will also produce different results.

**Comment by bmreiniger** [comment]
Thanks for the example!  (In the future, providing code as formatted text is more helpful: people can copy and paste to quickly retry what you're showing.)

This demonstrates what I was alluding to in my second paragraph: this is expected behavior.  If you print the results of the penultimate line (`fit_transform`), you'll see different values within each category.  You should get the same output if you change the last line to `cbe_encoder.transform(df1['f1'], df1['label'])`.  Using `fit_transform`, or `transform` with `y` specified, tells the package that you're transforming the training dataset, and so it takes the sliding transformation that CatBoost is known for.  On the other hand, when you `transform` with `y=None`, the package takes that as meaning you're transforming the test set, and so fixed values per category are used (roughly, the mean target from the entire training set).  See the NOTE at the end of the docstring.

**Comment by ccylance** [comment]
Very clear! Appreciate for your reply!

**Comment by ccylance** [comment]
I revisited the code and noticed that in the source code of CatBoost，the value is calculated by (coutinclass + piror)/(counts + priorDenominator) while in sklearn, the function is ( coutinclass - y + mean*a) /(counts + a), Why would there be a difference in this part?

**Comment by bmreiniger** [comment]
> @bmreiniger I revisited the code and noticed that in the source code of CatBoost，the value is calculated by (coutinclass + piror)/(counts + priorDenominator) while in sklearn, the function is ( coutinclass - y + mean*a) /(counts + a), Why would there be a difference in this part?

I'm not entirely sure, in particular what the CatBoost source's definitions of those terms are. But they seem to be just two (probably different but similar) ways to regularize/smooth the raw mean-target-so-far. The `mean` here is the global mean, a sensible default for a prior. And note that the `- y` part is just to remove the row's own contribution from pandas's cumulative sum.

**Comment by ccylance** [comment]
@bmreiniger  
When conducting a regression task, I observed that CatBoost implements bucketing for labels, which is not the case here. Is the bucketing process necessary in this context?

**Comment by bmreiniger** [comment]
I don't see why it would be, but maybe bucketing (depending on how you assign the label then) acts as another source of regularization?  Or maybe it's just faster?  Can you link to their source that performs bucketing?  (Maybe better to ask over there.)
